src/libcore/char.rs

```diff
@@ -654,3 +654,45 @@ impl Iterator for EncodeUtf16 {
         self.as_slice().iter().size_hint()
     }
 }
+
+
+/// An iterator over an iterator of bytes of the characters the bytes represent
+/// as UTF-8
+#[unstable(feature = "decode_utf8", issue = "33906")]
+#[derive(Clone, Debug)]
+pub struct DecodeUtf8<I: Iterator<Item = u8>>(I);
+
+/// Decodes an `Iterator` of bytes as UTF-8.
+#[unstable(feature = "decode_utf8", issue = "33906")]
+#[inline]
+pub fn decode_utf8<I: Iterator<Item = u8>>(i: I) -> DecodeUtf8<I> {
+    DecodeUtf8(i)
+}
+
+/// `<DecodeUtf8 as Iterator>::next` returns this for an invalid input sequence.
+#[unstable(feature = "decode_utf8", issue = "33906")]
+#[derive(Debug)]
+pub struct InvalidSequence;
```

Comment:
Could this be `struct InvalidSequence(())` so we can backwards-compatibly add more variants? You may want to also derive `PartialEq` for it so `assert_eq!` can work.